presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryMetadata.java

```diff
@@ -211,8 +234,11 @@ public synchronized MemoryInsertTableHandle beginInsert(ConnectorSession session
     {
         requireNonNull(handle, "handle is null");
         checkArgument(handle instanceof MemoryTableHandle);
+        MemoryTableHandle memoryTableHandle = (MemoryTableHandle) handle;
 
-        MemoryTableLayoutHandle layoutHandle = new MemoryTableLayoutHandle((MemoryTableHandle) handle);
+        MemoryTableLayoutHandle layoutHandle = new MemoryTableLayoutHandle(
+                memoryTableHandle,
+                ImmutableMap.copyOf(hostsPerTable.get((memoryTableHandle).getTableId())));
```

Comment:
`hostsPerTable.get` might be `null` if `beginCreateTable` but `finishInsert` didn't. There is a race condition here I think